src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -462,28 +490,462 @@ void Frame::denoise(
+            reader.read_tile(tile_x, tile_y, &shading_tile);
+
+            // No need to read beauty and filtered AOVs because they
+            // are in the shading buffer.
+
+            // Read unfiltered AOV layers.
+            vector<unique_ptr<Tile>> aov_tiles;
+            vector<const float*> aov_ptrs;
+            for (size_t i = 0; i < aovs().size(); ++i)
+            {
+                UnfilteredAOV* aov = dynamic_cast<UnfilteredAOV*>(
+                    aovs().get_by_index(i));
+
+                if (aov == nullptr)
+                    continue;
+
+                const string aov_name = aov->get_name();
+
+                // Search layer index in the file.
+                size_t subimage_index(~0);
+                for (size_t s = 0; s < checkpoint_props.size(); ++s)
+                {
+                    if (get<0>(checkpoint_props[s]) == aov_name)
+                    {
+                        subimage_index = s;
+                        break;
+                    }
+                }
+
+                assert(subimage_index != size_t(~0));
+
+                Image& aov_image = aov->get_image();
+                Tile& aov_tile = aov_image.tile(tile_x, tile_y);
+                reader.choose_subimage(subimage_index);
+                reader.read_tile(tile_x, tile_y, &aov_tile);
+            }
+        }
+    }
+
+    // Load internal AOVs (from external files).
+    for (size_t i = 0, e = internal_aovs().size(); i < e; ++i)
+    {
+        AOV* aov = internal_aovs().get_by_index(i);
+
+        DenoiserAOV* denoiser_aov = dynamic_cast<DenoiserAOV*>(aov);
+
+        // Load denoiser checkpoint.
+        if (denoiser_aov != nullptr)
+        {
+            if (!load_denoiser_checkpoint(impl->m_checkpoint_resume_path, denoiser_aov))
+                return false;
+        }
+    }
+
+    RENDERER_LOG_INFO("read checkpoint file %s, resuming rendering at pass %s.",
+        impl->m_checkpoint_resume_path.c_str(),
+        pretty_uint(start_pass).c_str());
+
+    impl->m_initial_pass = start_pass;
+
+    return true;
+}
+
+void Frame::save_checkpoint(
+    PermanentShadingResultFrameBufferFactory*   buffer_factory,
+    const size_t                                pass) const
+{
+    if (!impl->m_checkpoint_create)
+        return;
+
+    create_parent_directories(impl->m_checkpoint_create_path.c_str());
+
+    GenericImageFileWriter writer(impl->m_checkpoint_create_path.c_str());
+
+    ImageAttributes image_attributes = ImageAttributes::create_default_attributes();
+    image_attributes.insert("appleseed:LastPass", pass);
+
+    // Create layers.
+    // Buffer containing pixels' weight.
+    ShadingBufferCanvas weight_canvas(*this, buffer_factory);
+
+    // Create channel names.
+    const size_t shading_channel_count = weight_canvas.properties().m_channel_count;
+    vector<string> shading_channel_names; // for shading buffer
+    vector<const char *> shading_channel_names_cr;
```

Comment:
What does the `_cr` suffix mean here? What about `_cstr`?

Reply:
It was for `char`. but `_cstr` is better indeed